src/Symfony/Bundle/FrameworkBundle/Test/KernelTestCase.php

```diff
@@ -115,23 +127,59 @@ protected static function getKernelClass()
                     $dir = "$phpUnitDir/$dir";
                 }
             }
-        } else {
-            $dir = static::getPhpUnitXmlDir();
+
+            if ($kernelClass = static::findKernelClassInDirectory($dir)) {
+                throw new \RuntimeException(sprintf('There is no test kernel class in specified KERNEL_DIR "%s"', $_SERVER['KERNEL_CLASS']));
+            }
+
+            return $kernelClass;
         }
 
+        // try to find kernel class near phpunit.xml/phpunit.xml.dist files
+        if ($kernelClass = static::findKernelClassInDirectory(static::getPhpUnitXmlDir())) {
+            return $kernelClass;
+        }
+
+        // fallback to default kernel class
+        return static::getDefaultTestKernelClass();
+    }
+
+    /**
+     * Tries to find kernel class in directory. Returns null if no appropriative class found.
+     *
+     * @param string $dir directory to be processed
+     *
+     * @return string|null
+     */
+    protected static function findKernelClassInDirectory($dir)
+    {
+        $kernelClass = null;
+
         $finder = new Finder();
         $finder->name('*Kernel.php')->depth(0)->in($dir);
         $results = iterator_to_array($finder);
-        if (!count($results)) {
-            throw new \RuntimeException('Either set KERNEL_DIR in your phpunit.xml according to https://symfony.com/doc/current/book/testing.html#your-first-functional-test or override the WebTestCase::createKernel() method.');
-        }
+        if (count($results)) {
+            $file = current($results);
 
-        $file = current($results);
-        $class = $file->getBasename('.php');
+            $classes = get_declared_classes();
+            require_once $file;
+            $newClasses = array_diff(get_declared_classes(), $classes);
```

Comment:
this logic is broken if the file was already loaded previously, as it won't declare a new class

Reply:
Replaced with ClassFinder invocation